### ISSUE TITLE
Add LoDash Typings

### ISSUE THREAD

**Issue by bczengel** [bczengel/DefinitelyTyped#1]
*Use http://lodash.com/docs as the guide for comments, typings and tests.
*Update Comments to use lodash docs descriptions
*Update parameter names to use lodash doc parameter names
*Add additional optional parameters where applicable including _.pluck and _.where style overloads.
*Include all examples from lodash docs in tests.
*Cast results of examples to desired type to be "better" tests

**Comment by bczengel** [comment]
Array methods complete

**Comment by bczengel** [comment]
Collection methods complete

**Comment by bczengel** [comment]
Function methods complete

**Comment by georgiosd** [comment]
Hm... Just forked the project and tried to see where we are but I'll admit to being a little lost. I haven't used underscore in a while and I don't know how much ground there is between the two still. 

I'm also not sure whether it's a good idea to declare _ as a module vs an interface - using an interface allows to specify the type of lodash variables instead of just the global _ (e.g. if you're injecting lodash in AngularJS). 
Plus, we'll definitely need an interface when we get to chained functions.

Thoughts?

**Comment by bczengel** [comment]
You'd be surprised how much lodash has in common with underscore.  For the most part lodash is just more.  There have only been a few methods that I found that were different in a way that would make them not work if you called them the way underscore wanted them to be called.  That's why I decided to start with underscore.  I've been using underscore's typings in place of any suitable lodash typings for a while now.

As for using the module vs. the interface, I'm not sure that it matters.  The way it is now, I'm pretty sure you could just you "typeof _" when you wanted the type and "_" when you wanted the module.  I use requirejs compilation and I would use this definition like this:

```
///<reference path="typings/lodash/lodash.d.ts" />
import _lodash = require('_');
var _ = <typeof _>_lodash;
```

The way it is now, also has the chaining methods covered.  The first two lines that say:

```
declare function _<T>(value: Array<T>): _<T>;
declare function _<T>(value: T): _<T>;
```

...declare "_" as a function for beginning a chain.

The lines:

```
declare module _ {...
```

are where the _.methods live.

Then the line:

```
declare class _<T> {...
```

are where all the _(...).methods live.

I think everything is covered.  It just requires that we find and update the differences, add the new methods, and update the comments and method signatures to match the lodash nomenclature.

Did I hit everything?

-Brian

**Comment by bczengel** [comment]
Object methods complete

**Comment by bczengel** [comment]
Utility methods complete

**Comment by bczengel** [comment]
properties completed

**Comment by bczengel** [comment]
Core chaining methods completed, core array methods(push, pop, concat, etc.) complete, lodash chainable wrapper methods "after"-"difference" complete.

**Comment by georgiosd** [comment]
I'm sure you are right regarding the similarities between underscore and lodash, I was just saying that because I'm not aware of them, I found it hard to work off the underscore base instead of just writing it from scratch (at least for the method definitions).

It looks like you've done major headway anyway so I guess my help is obsolete. :)

But as soon as you publish a nuget for this I'll try it on for size and report any issues!

PS: I wasn't aware you could do that with TS modules

**Comment by bczengel** [comment]
Yeah, I had a 4 day weekend so I decided to spend it doing this.  I did decide to use an interface for chainable signatures.  I'm basically doing all the chainable stuff from scratch because I didn't like how it was done in the underscore typings.  The things I did like about the underscore typings were its use of generics and the handy interfaces they created for Collection, List, Dictionaries and Iterators.  

You can grab it now and try it out.  The only thing missing is the chainable logic.  I don't think I'll publish a nuget for it since it looks like they have that automated once it gets added to the main repo.

**Comment by gregwebs** [comment]
This looks great. My only suggestion is to tighten up the type names. Rather than
`LoDash.LoDashArrayWrapper` something like `_.ArrayWrap` might be better.

**Comment by bczengel** [comment]
Yeah I will probably do something like that.  The way it is now, I can't
use _ for the function interface and the module without a name conflict.
 So maybe I'll make it LoDash.ArrayWrapper.  We'll see.  The problem I have
right now is that I want Arrays to also be able to chain like an object
since arrays are objects too but every time I try to have the array wrapper
extend the object wrapper I run into a road block.  I'm sure I'll think of
something.

Brian

**Comment by gregwebs** [comment]
Overall the priority should be that the types the user types the most should be the shortest. I am not sure if the user needs to type the function interface type: that seems like it will always be inferred.

**Comment by georgiosd** [comment]
@bczengel Can you create a test case with the problem? Perhaps we can help solve it.

**Comment by gregwebs** [comment]
This isn't really something you can write a test case for. However, you can look at the tests as example usage to help make the design decision.

**Comment by bczengel** [comment]
@gregwebs, 
So it sounds like, just to clarify, you would like to see the lodash function/object referenced with `_` and also the interfaces namespaced with an `_` module like this `_.ArrayWrapper` as opposed to `LoDash.ArrayWrapper`.  That makes sense, I just don't know how to do it.

I guess we could just remove the module all together and just name the interfaces with the namespace included: `_ArrayWrapper` instead of `_.ArrayWrapper`.  I just really like having the module.  I think I still prefer `LoDash.ArrayWrapper` to `_ArrayWrapper`.

Plus, I really can't see many referencing the interfaces directly that often.  Most will only ever need to say `var lodash = <LoDash.LoDashStatic>_.NoConflict();` doing a noConflict or something like that.  But even then, `_.noConflict()` returns a instance of `LoDash.LoDashStatic`.  The nature of lodash is that you start with a type you control and you end with a type you control.  It seems weird that someone would, for instance pass around an `LoDash.ArrayWrapper` instead of just passing around the array.  So I'm not too concerned, at this point, with being verbose in my naming style.  I'm referencing them in the tests just to make sure that the interface returns what I expect them to return.  You wouldn't normally need to do that.

That being said, if you think you can come up with a better naming style, feel free to fork and make the change.  I've already "reorganized" things like 4 times so running through it again to update what I have in my working copy to your naming style won't be a big deal especially since it would just be a find and replace operation.  Once you get something working though, just share with us so we can poke any holes in it before you go through the effort of doing the full rename. 

I'm really new to Git and GitHub.  Is there a way to do a code review?

Brian

**Comment by gregwebs** [comment]
sending a pull request back to the main repo is the best way to initiate code review.

**Comment by gregwebs** [comment]
I am sending some pull requests from issues that are showing up with using this against a code base.

I changed the 3rd line to `declare module _ {` and it seems to work. This is very useful because I do have a few places that use the lodash types.

**Comment by georgiosd** [comment]
I'll agree that I haven't found a reason to use the interface names. @gregwebs can you give us an example of where this is needed?

**Comment by bczengel** [comment]
Hey, if it works that's great!!  I could have sworn I tried it and it caused errors.  I may have made some changes since I tried last that made it possible.  I give it another go and see if it works for me.  Thanks!!

**Comment by gregwebs** [comment]
I use LoDash interfaces when I create a genric function that calls a LoDash function. @georgiosd further discussion is probably going to take place on my pull request #2 now.

**Comment by georgiosd** [comment]
Hey guys - how is this going?

Edit: just saw there's a lodash dir at the DefinitelyTyped project which I assume it's your work, well done!

**Comment by bczengel** [comment]
Go super busy @ work.  I'm going to finish it up over my holiday break.

**Comment by georgiosd** [comment]
Hm... may wanna check this out then: https://github.com/borisyankov/DefinitelyTyped/tree/master/lodash

**Comment by bczengel** [comment]
That's mine. We submitted some pull requests about a month ago.  It just
doesn't have all the chainable methods yet. It's actually had a few people
make fixes and add some more chainable methods.
